<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>FlowableDoOnEachTest.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">RxJava_2_x$Unnamed.exec</a> &gt; <a href="index.source.html" class="el_package">io.reactivex.internal.operators.flowable</a> &gt; <span class="el_source">FlowableDoOnEachTest.java</span></div><h1>FlowableDoOnEachTest.java</h1><pre class="source lang-java linenums">/**
 * Copyright (c) 2016-present, RxJava Contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package io.reactivex.internal.operators.flowable;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.*;
import org.reactivestreams.*;

import io.reactivex.*;
import io.reactivex.exceptions.*;
import io.reactivex.flowables.ConnectableFlowable;
import io.reactivex.functions.*;
import io.reactivex.internal.functions.Functions;
import io.reactivex.internal.fuseable.*;
import io.reactivex.internal.subscriptions.BooleanSubscription;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.processors.UnicastProcessor;
import io.reactivex.subscribers.*;

<span class="fc" id="L38">public class FlowableDoOnEachTest {</span>

    Subscriber&lt;String&gt; subscribedSubscriber;
    Subscriber&lt;String&gt; sideEffectSubscriber;

    @Before
    public void before() {
<span class="fc" id="L45">        subscribedSubscriber = TestHelper.mockSubscriber();</span>
<span class="fc" id="L46">        sideEffectSubscriber = TestHelper.mockSubscriber();</span>
<span class="fc" id="L47">    }</span>

    @Test
    public void testDoOnEach() {
<span class="fc" id="L51">        Flowable&lt;String&gt; base = Flowable.just(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;);</span>
<span class="fc" id="L52">        Flowable&lt;String&gt; doOnEach = base.doOnEach(sideEffectSubscriber);</span>

<span class="fc" id="L54">        doOnEach.subscribe(subscribedSubscriber);</span>

        // ensure the leaf observer is still getting called
<span class="fc" id="L57">        verify(subscribedSubscriber, never()).onError(any(Throwable.class));</span>
<span class="fc" id="L58">        verify(subscribedSubscriber, times(1)).onNext(&quot;a&quot;);</span>
<span class="fc" id="L59">        verify(subscribedSubscriber, times(1)).onNext(&quot;b&quot;);</span>
<span class="fc" id="L60">        verify(subscribedSubscriber, times(1)).onNext(&quot;c&quot;);</span>
<span class="fc" id="L61">        verify(subscribedSubscriber, times(1)).onComplete();</span>

        // ensure our injected observer is getting called
<span class="fc" id="L64">        verify(sideEffectSubscriber, never()).onError(any(Throwable.class));</span>
<span class="fc" id="L65">        verify(sideEffectSubscriber, times(1)).onNext(&quot;a&quot;);</span>
<span class="fc" id="L66">        verify(sideEffectSubscriber, times(1)).onNext(&quot;b&quot;);</span>
<span class="fc" id="L67">        verify(sideEffectSubscriber, times(1)).onNext(&quot;c&quot;);</span>
<span class="fc" id="L68">        verify(sideEffectSubscriber, times(1)).onComplete();</span>
<span class="fc" id="L69">    }</span>

    @Test
    public void testDoOnEachWithError() {
<span class="fc" id="L73">        Flowable&lt;String&gt; base = Flowable.just(&quot;one&quot;, &quot;fail&quot;, &quot;two&quot;, &quot;three&quot;, &quot;fail&quot;);</span>
<span class="fc" id="L74">        Flowable&lt;String&gt; errs = base.map(new Function&lt;String, String&gt;() {</span>
            @Override
            public String apply(String s) {
<span class="fc bfc" id="L77" title="All 2 branches covered.">                if (&quot;fail&quot;.equals(s)) {</span>
<span class="fc" id="L78">                    throw new RuntimeException(&quot;Forced Failure&quot;);</span>
                }
<span class="fc" id="L80">                return s;</span>
            }
        });

<span class="fc" id="L84">        Flowable&lt;String&gt; doOnEach = errs.doOnEach(sideEffectSubscriber);</span>

<span class="fc" id="L86">        doOnEach.subscribe(subscribedSubscriber);</span>
<span class="fc" id="L87">        verify(subscribedSubscriber, times(1)).onNext(&quot;one&quot;);</span>
<span class="fc" id="L88">        verify(subscribedSubscriber, never()).onNext(&quot;two&quot;);</span>
<span class="fc" id="L89">        verify(subscribedSubscriber, never()).onNext(&quot;three&quot;);</span>
<span class="fc" id="L90">        verify(subscribedSubscriber, never()).onComplete();</span>
<span class="fc" id="L91">        verify(subscribedSubscriber, times(1)).onError(any(Throwable.class));</span>

<span class="fc" id="L93">        verify(sideEffectSubscriber, times(1)).onNext(&quot;one&quot;);</span>
<span class="fc" id="L94">        verify(sideEffectSubscriber, never()).onNext(&quot;two&quot;);</span>
<span class="fc" id="L95">        verify(sideEffectSubscriber, never()).onNext(&quot;three&quot;);</span>
<span class="fc" id="L96">        verify(sideEffectSubscriber, never()).onComplete();</span>
<span class="fc" id="L97">        verify(sideEffectSubscriber, times(1)).onError(any(Throwable.class));</span>
<span class="fc" id="L98">    }</span>

    @Test
    public void testDoOnEachWithErrorInCallback() {
<span class="fc" id="L102">        Flowable&lt;String&gt; base = Flowable.just(&quot;one&quot;, &quot;two&quot;, &quot;fail&quot;, &quot;three&quot;);</span>
<span class="fc" id="L103">        Flowable&lt;String&gt; doOnEach = base.doOnNext(new Consumer&lt;String&gt;() {</span>
            @Override
            public void accept(String s) {
<span class="fc bfc" id="L106" title="All 2 branches covered.">                if (&quot;fail&quot;.equals(s)) {</span>
<span class="fc" id="L107">                    throw new RuntimeException(&quot;Forced Failure&quot;);</span>
                }
<span class="fc" id="L109">            }</span>
        });

<span class="fc" id="L112">        doOnEach.subscribe(subscribedSubscriber);</span>
<span class="fc" id="L113">        verify(subscribedSubscriber, times(1)).onNext(&quot;one&quot;);</span>
<span class="fc" id="L114">        verify(subscribedSubscriber, times(1)).onNext(&quot;two&quot;);</span>
<span class="fc" id="L115">        verify(subscribedSubscriber, never()).onNext(&quot;three&quot;);</span>
<span class="fc" id="L116">        verify(subscribedSubscriber, never()).onComplete();</span>
<span class="fc" id="L117">        verify(subscribedSubscriber, times(1)).onError(any(Throwable.class));</span>

<span class="fc" id="L119">    }</span>

    @Test
    public void testIssue1451Case1() {
        // https://github.com/Netflix/RxJava/issues/1451
<span class="fc" id="L124">        final int expectedCount = 3;</span>
<span class="fc" id="L125">        final AtomicInteger count = new AtomicInteger();</span>
<span class="fc bfc" id="L126" title="All 2 branches covered.">        for (int i = 0; i &lt; expectedCount; i++) {</span>
<span class="fc" id="L127">            Flowable</span>
<span class="fc" id="L128">                    .just(Boolean.TRUE, Boolean.FALSE)</span>
<span class="fc" id="L129">                    .takeWhile(new Predicate&lt;Boolean&gt;() {</span>
                        @Override
                        public boolean test(Boolean value) {
<span class="fc" id="L132">                            return value;</span>
                        }
                    })
<span class="fc" id="L135">                    .toList()</span>
<span class="fc" id="L136">                    .doOnSuccess(new Consumer&lt;List&lt;Boolean&gt;&gt;() {</span>
                        @Override
                        public void accept(List&lt;Boolean&gt; booleans) {
<span class="fc" id="L139">                            count.incrementAndGet();</span>
<span class="fc" id="L140">                        }</span>
                    })
<span class="fc" id="L142">                    .subscribe();</span>
        }
<span class="fc" id="L144">        assertEquals(expectedCount, count.get());</span>
<span class="fc" id="L145">    }</span>

    @Test
    public void testIssue1451Case2() {
        // https://github.com/Netflix/RxJava/issues/1451
<span class="fc" id="L150">        final int expectedCount = 3;</span>
<span class="fc" id="L151">        final AtomicInteger count = new AtomicInteger();</span>
<span class="fc bfc" id="L152" title="All 2 branches covered.">        for (int i = 0; i &lt; expectedCount; i++) {</span>
<span class="fc" id="L153">            Flowable</span>
<span class="fc" id="L154">                    .just(Boolean.TRUE, Boolean.FALSE, Boolean.FALSE)</span>
<span class="fc" id="L155">                    .takeWhile(new Predicate&lt;Boolean&gt;() {</span>
                        @Override
                        public boolean test(Boolean value) {
<span class="fc" id="L158">                            return value;</span>
                        }
                    })
<span class="fc" id="L161">                    .toList()</span>
<span class="fc" id="L162">                    .doOnSuccess(new Consumer&lt;List&lt;Boolean&gt;&gt;() {</span>
                        @Override
                        public void accept(List&lt;Boolean&gt; booleans) {
<span class="fc" id="L165">                            count.incrementAndGet();</span>
<span class="fc" id="L166">                        }</span>
                    })
<span class="fc" id="L168">                    .subscribe();</span>
        }
<span class="fc" id="L170">        assertEquals(expectedCount, count.get());</span>
<span class="fc" id="L171">    }</span>

    @Test
    @Ignore(&quot;crashing publisher can't propagate to a subscriber&quot;)
    public void testFatalError() {
//        try {
//            Flowable.just(1, 2, 3)
//                    .flatMap(new Function&lt;Integer, Flowable&lt;?&gt;&gt;() {
//                        @Override
//                        public Flowable&lt;?&gt; apply(Integer integer) {
//                            return Flowable.create(new Publisher&lt;Object&gt;() {
//                                @Override
//                                public void subscribe(Subscriber&lt;Object&gt; subscriber) {
//                                    throw new NullPointerException(&quot;Test NPE&quot;);
//                                }
//                            });
//                        }
//                    })
//                    .doOnNext(new Consumer&lt;Object&gt;() {
//                        @Override
//                        public void accept(Object o) {
//                            System.out.println(&quot;Won't come here&quot;);
//                        }
//                    })
//                    .subscribe();
//            fail(&quot;should have thrown an exception&quot;);
//        } catch (OnErrorNotImplementedException e) {
//            assertTrue(e.getCause() instanceof NullPointerException);
//            assertEquals(e.getCause().getMessage(), &quot;Test NPE&quot;);
//            System.out.println(&quot;Received exception: &quot; + e);
//        }
<span class="nc" id="L202">    }</span>

    @Test
    public void onErrorThrows() {
<span class="fc" id="L206">        TestSubscriber&lt;Object&gt; ts = TestSubscriber.create();</span>

<span class="fc" id="L208">        Flowable.error(new TestException())</span>
<span class="fc" id="L209">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) {
<span class="fc" id="L212">                throw new TestException();</span>
            }
<span class="fc" id="L214">        }).subscribe(ts);</span>

<span class="fc" id="L216">        ts.assertNoValues();</span>
<span class="fc" id="L217">        ts.assertNotComplete();</span>
<span class="fc" id="L218">        ts.assertError(CompositeException.class);</span>

<span class="fc" id="L220">        CompositeException ex = (CompositeException)ts.errors().get(0);</span>

<span class="fc" id="L222">        List&lt;Throwable&gt; exceptions = ex.getExceptions();</span>
<span class="fc" id="L223">        assertEquals(2, exceptions.size());</span>
<span class="fc" id="L224">        Assert.assertTrue(exceptions.get(0) instanceof TestException);</span>
<span class="fc" id="L225">        Assert.assertTrue(exceptions.get(1) instanceof TestException);</span>
<span class="fc" id="L226">    }</span>

    @Test
    public void ignoreCancel() {
<span class="fc" id="L230">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L233">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L236">                    s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L237">                    s.onNext(1);</span>
<span class="fc" id="L238">                    s.onNext(2);</span>
<span class="fc" id="L239">                    s.onError(new IOException());</span>
<span class="fc" id="L240">                    s.onComplete();</span>
<span class="fc" id="L241">                }</span>
            })
<span class="fc" id="L243">            .doOnNext(new Consumer&lt;Object&gt;() {</span>
                @Override
                public void accept(Object e) throws Exception {
<span class="fc" id="L246">                    throw new TestException();</span>
                }
            })
<span class="fc" id="L249">            .test()</span>
<span class="fc" id="L250">            .assertFailure(TestException.class);</span>

<span class="fc" id="L252">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L254">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L256">    }</span>

    @Test
    public void onErrorAfterCrash() {
<span class="fc" id="L260">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L263">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L266">                    s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L267">                    s.onError(new TestException());</span>
<span class="fc" id="L268">                }</span>
            })
<span class="fc" id="L270">            .doAfterTerminate(new Action() {</span>
                @Override
                public void run() throws Exception {
<span class="fc" id="L273">                    throw new IOException();</span>
                }
            })
<span class="fc" id="L276">            .test()</span>
<span class="fc" id="L277">            .assertFailure(TestException.class);</span>

<span class="fc" id="L279">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L281">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L283">    }</span>

    @Test
    public void onCompleteAfterCrash() {
<span class="fc" id="L287">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L290">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L293">                    s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L294">                    s.onComplete();</span>
<span class="fc" id="L295">                }</span>
            })
<span class="fc" id="L297">            .doAfterTerminate(new Action() {</span>
                @Override
                public void run() throws Exception {
<span class="fc" id="L300">                    throw new IOException();</span>
                }
            })
<span class="fc" id="L303">            .test()</span>
<span class="fc" id="L304">            .assertResult();</span>

<span class="fc" id="L306">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L308">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L310">    }</span>

    @Test
    public void onCompleteCrash() {
<span class="fc" id="L314">        Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
            @Override
            public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L317">                s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L318">                s.onComplete();</span>
<span class="fc" id="L319">            }</span>
        })
<span class="fc" id="L321">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L324">                throw new IOException();</span>
            }
        })
<span class="fc" id="L327">        .test()</span>
<span class="fc" id="L328">        .assertFailure(IOException.class);</span>
<span class="fc" id="L329">    }</span>

    @Test
    public void ignoreCancelConditional() {
<span class="fc" id="L333">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L336">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L339">                    s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L340">                    s.onNext(1);</span>
<span class="fc" id="L341">                    s.onNext(2);</span>
<span class="fc" id="L342">                    s.onError(new IOException());</span>
<span class="fc" id="L343">                    s.onComplete();</span>
<span class="fc" id="L344">                }</span>
            })
<span class="fc" id="L346">            .doOnNext(new Consumer&lt;Object&gt;() {</span>
                @Override
                public void accept(Object e) throws Exception {
<span class="fc" id="L349">                    throw new TestException();</span>
                }
            })
<span class="fc" id="L352">            .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L353">            .test()</span>
<span class="fc" id="L354">            .assertFailure(TestException.class);</span>

<span class="fc" id="L356">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L358">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L360">    }</span>

    @Test
    public void ignoreCancelConditional2() {
<span class="fc" id="L364">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L367">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L370">                    ConditionalSubscriber&lt;? super Object&gt; cs = (ConditionalSubscriber&lt;? super Object&gt;)s;</span>
<span class="fc" id="L371">                    cs.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L372">                    cs.tryOnNext(1);</span>
<span class="fc" id="L373">                    cs.tryOnNext(2);</span>
<span class="fc" id="L374">                    cs.onError(new IOException());</span>
<span class="fc" id="L375">                    cs.onComplete();</span>
<span class="fc" id="L376">                }</span>
            })
<span class="fc" id="L378">            .doOnNext(new Consumer&lt;Object&gt;() {</span>
                @Override
                public void accept(Object e) throws Exception {
<span class="fc" id="L381">                    throw new TestException();</span>
                }
            })
<span class="fc" id="L384">            .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L385">            .test()</span>
<span class="fc" id="L386">            .assertFailure(TestException.class);</span>

<span class="fc" id="L388">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L390">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L392">    }</span>

    @Test
    public void onErrorAfterCrashConditional() {
<span class="fc" id="L396">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L399">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L402">                    s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L403">                    s.onError(new TestException());</span>
<span class="fc" id="L404">                }</span>
            })
<span class="fc" id="L406">            .doAfterTerminate(new Action() {</span>
                @Override
                public void run() throws Exception {
<span class="fc" id="L409">                    throw new IOException();</span>
                }
            })
<span class="fc" id="L412">            .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L413">            .test()</span>
<span class="fc" id="L414">            .assertFailure(TestException.class);</span>

<span class="fc" id="L416">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L418">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L420">    }</span>

    @Test
    public void onCompleteAfter() {
<span class="fc" id="L424">        final int[] call = { 0 };</span>
<span class="fc" id="L425">        Flowable.just(1)</span>
<span class="fc" id="L426">        .doAfterTerminate(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L429">                call[0]++;</span>
<span class="fc" id="L430">            }</span>
        })
<span class="fc" id="L432">        .test()</span>
<span class="fc" id="L433">        .assertResult(1);</span>

<span class="fc" id="L435">        assertEquals(1, call[0]);</span>
<span class="fc" id="L436">    }</span>

    @Test
    public void onCompleteAfterCrashConditional() {
<span class="fc" id="L440">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L443">            Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
                @Override
                public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L446">                    s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L447">                    s.onComplete();</span>
<span class="fc" id="L448">                }</span>
            })
<span class="fc" id="L450">            .doAfterTerminate(new Action() {</span>
                @Override
                public void run() throws Exception {
<span class="fc" id="L453">                    throw new IOException();</span>
                }
            })
<span class="fc" id="L456">            .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L457">            .test()</span>
<span class="fc" id="L458">            .assertResult();</span>

<span class="fc" id="L460">            TestHelper.assertUndeliverable(errors, 0, IOException.class);</span>
        } finally {
<span class="fc" id="L462">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L464">    }</span>

    @Test
    public void onCompleteCrashConditional() {
<span class="fc" id="L468">        Flowable.fromPublisher(new Publisher&lt;Object&gt;() {</span>
            @Override
            public void subscribe(Subscriber&lt;? super Object&gt; s) {
<span class="fc" id="L471">                s.onSubscribe(new BooleanSubscription());</span>
<span class="fc" id="L472">                s.onComplete();</span>
<span class="fc" id="L473">            }</span>
        })
<span class="fc" id="L475">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L478">                throw new IOException();</span>
            }
        })
<span class="fc" id="L481">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L482">        .test()</span>
<span class="fc" id="L483">        .assertFailure(IOException.class);</span>
<span class="fc" id="L484">    }</span>

    @Test
    public void onErrorOnErrorCrashConditional() {
<span class="fc" id="L488">        TestSubscriber&lt;Object&gt; ts = Flowable.error(new TestException(&quot;Outer&quot;))</span>
<span class="fc" id="L489">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L492">                throw new TestException(&quot;Inner&quot;);</span>
            }
        })
<span class="fc" id="L495">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L496">        .test()</span>
<span class="fc" id="L497">        .assertFailure(CompositeException.class);</span>

<span class="fc" id="L499">        List&lt;Throwable&gt; errors = TestHelper.compositeList(ts.errors().get(0));</span>

<span class="fc" id="L501">        TestHelper.assertError(errors, 0, TestException.class, &quot;Outer&quot;);</span>
<span class="fc" id="L502">        TestHelper.assertError(errors, 1, TestException.class, &quot;Inner&quot;);</span>
<span class="fc" id="L503">    }</span>

    @Test
    public void fused() {
<span class="fc" id="L507">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L509">        final int[] call = { 0, 0 };</span>

<span class="fc" id="L511">        Flowable.range(1, 5)</span>
<span class="fc" id="L512">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L515">                call[0]++;</span>
<span class="fc" id="L516">            }</span>
        })
<span class="fc" id="L518">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L521">                call[1]++;</span>
<span class="fc" id="L522">            }</span>
        })
<span class="fc" id="L524">        .subscribe(ts);</span>

<span class="fc" id="L526">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L527">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.SYNC))</span>
<span class="fc" id="L528">        .assertResult(1, 2, 3, 4, 5);</span>

<span class="fc" id="L530">        assertEquals(5, call[0]);</span>
<span class="fc" id="L531">        assertEquals(1, call[1]);</span>
<span class="fc" id="L532">    }</span>

    @Test
    public void fusedOnErrorCrash() {
<span class="fc" id="L536">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L538">        final int[] call = { 0 };</span>

<span class="fc" id="L540">        Flowable.range(1, 5)</span>
<span class="fc" id="L541">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L544">                throw new TestException();</span>
            }
        })
<span class="fc" id="L547">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="nc" id="L550">                call[0]++;</span>
<span class="nc" id="L551">            }</span>
        })
<span class="fc" id="L553">        .subscribe(ts);</span>

<span class="fc" id="L555">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L556">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.SYNC))</span>
<span class="fc" id="L557">        .assertFailure(TestException.class);</span>

<span class="fc" id="L559">        assertEquals(0, call[0]);</span>
<span class="fc" id="L560">    }</span>

    @Test
    public void fusedConditional() {
<span class="fc" id="L564">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L566">        final int[] call = { 0, 0 };</span>

<span class="fc" id="L568">        Flowable.range(1, 5)</span>
<span class="fc" id="L569">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L572">                call[0]++;</span>
<span class="fc" id="L573">            }</span>
        })
<span class="fc" id="L575">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L578">                call[1]++;</span>
<span class="fc" id="L579">            }</span>
        })
<span class="fc" id="L581">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L582">        .subscribe(ts);</span>

<span class="fc" id="L584">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L585">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.SYNC))</span>
<span class="fc" id="L586">        .assertResult(1, 2, 3, 4, 5);</span>

<span class="fc" id="L588">        assertEquals(5, call[0]);</span>
<span class="fc" id="L589">        assertEquals(1, call[1]);</span>
<span class="fc" id="L590">    }</span>

    @Test
    public void fusedOnErrorCrashConditional() {
<span class="fc" id="L594">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L596">        final int[] call = { 0 };</span>

<span class="fc" id="L598">        Flowable.range(1, 5)</span>
<span class="fc" id="L599">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L602">                throw new TestException();</span>
            }
        })
<span class="fc" id="L605">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="nc" id="L608">                call[0]++;</span>
<span class="nc" id="L609">            }</span>
        })
<span class="fc" id="L611">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L612">        .subscribe(ts);</span>

<span class="fc" id="L614">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L615">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.SYNC))</span>
<span class="fc" id="L616">        .assertFailure(TestException.class);</span>

<span class="fc" id="L618">        assertEquals(0, call[0]);</span>
<span class="fc" id="L619">    }</span>

    @Test
    public void fusedAsync() {
<span class="fc" id="L623">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L625">        final int[] call = { 0, 0 };</span>

<span class="fc" id="L627">        UnicastProcessor&lt;Integer&gt; up = UnicastProcessor.create();</span>

<span class="fc" id="L629">        up</span>
<span class="fc" id="L630">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L633">                call[0]++;</span>
<span class="fc" id="L634">            }</span>
        })
<span class="fc" id="L636">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L639">                call[1]++;</span>
<span class="fc" id="L640">            }</span>
        })
<span class="fc" id="L642">        .subscribe(ts);</span>

<span class="fc" id="L644">        TestHelper.emit(up, 1, 2, 3, 4, 5);</span>

<span class="fc" id="L646">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L647">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.ASYNC))</span>
<span class="fc" id="L648">        .assertResult(1, 2, 3, 4, 5);</span>

<span class="fc" id="L650">        assertEquals(5, call[0]);</span>
<span class="fc" id="L651">        assertEquals(1, call[1]);</span>
<span class="fc" id="L652">    }</span>

    @Test
    public void fusedAsyncConditional() {
<span class="fc" id="L656">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L658">        final int[] call = { 0, 0 };</span>

<span class="fc" id="L660">        UnicastProcessor&lt;Integer&gt; up = UnicastProcessor.create();</span>

<span class="fc" id="L662">        up</span>
<span class="fc" id="L663">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L666">                call[0]++;</span>
<span class="fc" id="L667">            }</span>
        })
<span class="fc" id="L669">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L672">                call[1]++;</span>
<span class="fc" id="L673">            }</span>
        })
<span class="fc" id="L675">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L676">        .subscribe(ts);</span>

<span class="fc" id="L678">        TestHelper.emit(up, 1, 2, 3, 4, 5);</span>

<span class="fc" id="L680">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L681">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.ASYNC))</span>
<span class="fc" id="L682">        .assertResult(1, 2, 3, 4, 5);</span>

<span class="fc" id="L684">        assertEquals(5, call[0]);</span>
<span class="fc" id="L685">        assertEquals(1, call[1]);</span>
<span class="fc" id="L686">    }</span>

    @Test
    public void fusedAsyncConditional2() {
<span class="fc" id="L690">        TestSubscriber&lt;Integer&gt; ts = SubscriberFusion.newTest(QueueFuseable.ANY);</span>

<span class="fc" id="L692">        final int[] call = { 0, 0 };</span>

<span class="fc" id="L694">        UnicastProcessor&lt;Integer&gt; up = UnicastProcessor.create();</span>

<span class="fc" id="L696">        up.hide()</span>
<span class="fc" id="L697">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L700">                call[0]++;</span>
<span class="fc" id="L701">            }</span>
        })
<span class="fc" id="L703">        .doOnComplete(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="fc" id="L706">                call[1]++;</span>
<span class="fc" id="L707">            }</span>
        })
<span class="fc" id="L709">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L710">        .subscribe(ts);</span>

<span class="fc" id="L712">        TestHelper.emit(up, 1, 2, 3, 4, 5);</span>

<span class="fc" id="L714">        ts.assertOf(SubscriberFusion.&lt;Integer&gt;assertFuseable())</span>
<span class="fc" id="L715">        .assertOf(SubscriberFusion.&lt;Integer&gt;assertFusionMode(QueueFuseable.NONE))</span>
<span class="fc" id="L716">        .assertResult(1, 2, 3, 4, 5);</span>

<span class="fc" id="L718">        assertEquals(5, call[0]);</span>
<span class="fc" id="L719">        assertEquals(1, call[1]);</span>
<span class="fc" id="L720">    }</span>

    @Test
    public void dispose() {
<span class="fc" id="L724">        TestHelper.checkDisposed(Flowable.just(1).doOnEach(new TestSubscriber&lt;Integer&gt;()));</span>
<span class="fc" id="L725">    }</span>

    @Test
    public void doubleOnSubscribe() {
<span class="fc" id="L729">        TestHelper.checkDoubleOnSubscribeFlowable(new Function&lt;Flowable&lt;Object&gt;, Flowable&lt;Object&gt;&gt;() {</span>
            @Override
            public Flowable&lt;Object&gt; apply(Flowable&lt;Object&gt; f) throws Exception {
<span class="fc" id="L732">                return f.doOnEach(new TestSubscriber&lt;Object&gt;());</span>
            }
        });
<span class="fc" id="L735">    }</span>

    @Test
    public void doOnNextDoOnErrorFused() {
<span class="fc" id="L739">        ConnectableFlowable&lt;Integer&gt; cf = Flowable.just(1)</span>
<span class="fc" id="L740">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L743">                throw new TestException(&quot;First&quot;);</span>
            }
        })
<span class="fc" id="L746">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L749">                throw new TestException(&quot;Second&quot;);</span>
            }
        })
<span class="fc" id="L752">        .publish();</span>

<span class="fc" id="L754">        TestSubscriber&lt;Integer&gt; ts = cf.test();</span>
<span class="fc" id="L755">        cf.connect();</span>

<span class="fc" id="L757">        ts.assertFailure(CompositeException.class);</span>

<span class="fc" id="L759">        TestHelper.assertError(ts, 0, TestException.class, &quot;First&quot;);</span>
<span class="fc" id="L760">        TestHelper.assertError(ts, 1, TestException.class, &quot;Second&quot;);</span>
<span class="fc" id="L761">    }</span>

    @Test
    public void doOnNextDoOnErrorCombinedFused() {
<span class="fc" id="L765">        ConnectableFlowable&lt;Integer&gt; cf = Flowable.just(1)</span>
<span class="fc" id="L766">                .compose(new FlowableTransformer&lt;Integer, Integer&gt;() {</span>
                    @Override
                    public Publisher&lt;Integer&gt; apply(Flowable&lt;Integer&gt; v) {
<span class="fc" id="L769">                        return new FlowableDoOnEach&lt;Integer&gt;(v,</span>
<span class="fc" id="L770">                                new Consumer&lt;Integer&gt;() {</span>
                                    @Override
                                    public void accept(Integer v) throws Exception {
<span class="fc" id="L773">                                        throw new TestException(&quot;First&quot;);</span>
                                    }
                                },
<span class="fc" id="L776">                                new Consumer&lt;Throwable&gt;() {</span>
                                    @Override
                                    public void accept(Throwable e) throws Exception {
<span class="fc" id="L779">                                        throw new TestException(&quot;Second&quot;);</span>
                                    }
                                },
                                Functions.EMPTY_ACTION
                                ,
                                Functions.EMPTY_ACTION
                                );
                    }
                })
<span class="fc" id="L788">        .publish();</span>

<span class="fc" id="L790">        TestSubscriber&lt;Integer&gt; ts = cf.test();</span>
<span class="fc" id="L791">        cf.connect();</span>

<span class="fc" id="L793">        ts.assertFailure(CompositeException.class);</span>

<span class="fc" id="L795">        TestHelper.assertError(ts, 0, TestException.class, &quot;First&quot;);</span>
<span class="fc" id="L796">        TestHelper.assertError(ts, 1, TestException.class, &quot;Second&quot;);</span>
<span class="fc" id="L797">    }</span>

    @Test
    public void doOnNextDoOnErrorFused2() {
<span class="fc" id="L801">        ConnectableFlowable&lt;Integer&gt; cf = Flowable.just(1)</span>
<span class="fc" id="L802">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L805">                throw new TestException(&quot;First&quot;);</span>
            }
        })
<span class="fc" id="L808">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L811">                throw new TestException(&quot;Second&quot;);</span>
            }
        })
<span class="fc" id="L814">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L817">                throw new TestException(&quot;Third&quot;);</span>
            }
        })
<span class="fc" id="L820">        .publish();</span>

<span class="fc" id="L822">        TestSubscriber&lt;Integer&gt; ts = cf.test();</span>
<span class="fc" id="L823">        cf.connect();</span>

<span class="fc" id="L825">        ts.assertFailure(CompositeException.class);</span>

<span class="fc" id="L827">        TestHelper.assertError(ts, 0, TestException.class, &quot;First&quot;);</span>
<span class="fc" id="L828">        TestHelper.assertError(ts, 1, TestException.class, &quot;Second&quot;);</span>
<span class="fc" id="L829">        TestHelper.assertError(ts, 2, TestException.class, &quot;Third&quot;);</span>
<span class="fc" id="L830">    }</span>

    @Test
    public void doOnNextDoOnErrorFusedConditional() {
<span class="fc" id="L834">        ConnectableFlowable&lt;Integer&gt; cf = Flowable.just(1)</span>
<span class="fc" id="L835">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L838">                throw new TestException(&quot;First&quot;);</span>
            }
        })
<span class="fc" id="L841">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L844">                throw new TestException(&quot;Second&quot;);</span>
            }
        })
<span class="fc" id="L847">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L848">        .publish();</span>

<span class="fc" id="L850">        TestSubscriber&lt;Integer&gt; ts = cf.test();</span>
<span class="fc" id="L851">        cf.connect();</span>

<span class="fc" id="L853">        ts.assertFailure(CompositeException.class);</span>

<span class="fc" id="L855">        TestHelper.assertError(ts, 0, TestException.class, &quot;First&quot;);</span>
<span class="fc" id="L856">        TestHelper.assertError(ts, 1, TestException.class, &quot;Second&quot;);</span>
<span class="fc" id="L857">    }</span>

    @Test
    public void doOnNextDoOnErrorFusedConditional2() {
<span class="fc" id="L861">        ConnectableFlowable&lt;Integer&gt; cf = Flowable.just(1)</span>
<span class="fc" id="L862">        .doOnNext(new Consumer&lt;Integer&gt;() {</span>
            @Override
            public void accept(Integer v) throws Exception {
<span class="fc" id="L865">                throw new TestException(&quot;First&quot;);</span>
            }
        })
<span class="fc" id="L868">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L871">                throw new TestException(&quot;Second&quot;);</span>
            }
        })
<span class="fc" id="L874">        .doOnError(new Consumer&lt;Throwable&gt;() {</span>
            @Override
            public void accept(Throwable e) throws Exception {
<span class="fc" id="L877">                throw new TestException(&quot;Third&quot;);</span>
            }
        })
<span class="fc" id="L880">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L881">        .publish();</span>

<span class="fc" id="L883">        TestSubscriber&lt;Integer&gt; ts = cf.test();</span>
<span class="fc" id="L884">        cf.connect();</span>

<span class="fc" id="L886">        ts.assertFailure(CompositeException.class);</span>

<span class="fc" id="L888">        TestHelper.assertError(ts, 0, TestException.class, &quot;First&quot;);</span>
<span class="fc" id="L889">        TestHelper.assertError(ts, 1, TestException.class, &quot;Second&quot;);</span>
<span class="fc" id="L890">        TestHelper.assertError(ts, 2, TestException.class, &quot;Third&quot;);</span>
<span class="fc" id="L891">    }</span>

    @Test
    public void doOnNextDoOnErrorCombinedFusedConditional() {
<span class="fc" id="L895">        ConnectableFlowable&lt;Integer&gt; cf = Flowable.just(1)</span>
<span class="fc" id="L896">                .compose(new FlowableTransformer&lt;Integer, Integer&gt;() {</span>
                    @Override
                    public Publisher&lt;Integer&gt; apply(Flowable&lt;Integer&gt; v) {
<span class="fc" id="L899">                        return new FlowableDoOnEach&lt;Integer&gt;(v,</span>
<span class="fc" id="L900">                                new Consumer&lt;Integer&gt;() {</span>
                                    @Override
                                    public void accept(Integer v) throws Exception {
<span class="fc" id="L903">                                        throw new TestException(&quot;First&quot;);</span>
                                    }
                                },
<span class="fc" id="L906">                                new Consumer&lt;Throwable&gt;() {</span>
                                    @Override
                                    public void accept(Throwable e) throws Exception {
<span class="fc" id="L909">                                        throw new TestException(&quot;Second&quot;);</span>
                                    }
                                },
                                Functions.EMPTY_ACTION
                                ,
                                Functions.EMPTY_ACTION
                                );
                    }
                })
<span class="fc" id="L918">        .filter(Functions.alwaysTrue())</span>
<span class="fc" id="L919">        .publish();</span>

<span class="fc" id="L921">        TestSubscriber&lt;Integer&gt; ts = cf.test();</span>
<span class="fc" id="L922">        cf.connect();</span>

<span class="fc" id="L924">        ts.assertFailure(CompositeException.class);</span>

<span class="fc" id="L926">        TestHelper.assertError(ts, 0, TestException.class, &quot;First&quot;);</span>
<span class="fc" id="L927">        TestHelper.assertError(ts, 1, TestException.class, &quot;Second&quot;);</span>
<span class="fc" id="L928">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.2.201808211720</span></div></body></html>